#6041: CCL refactoring and sharded allgather optimization

SeanNijjar opened the pull request:
Now that FD2 has dropped, and with it a lot of stability improvements, I'm going through my backlog of CCL changes and starting to merge them into main.

Unfortunately, due to the backlog, some of my changes ended up being intertwined and weren't easily  separable. For that reason, reason this PR needed to include some substantial refactor work in preparation for the reduce scatter op. Those refactors are included alongside some sharded allgather optimizations and bug fixes.

At this point, the sharded allgather supports single-tile-high shards only. Multi-tile-high sharded support are coming in a future change.

@tt-rkim - FYI adding myself as codeowner for new CCL directory
@tt-aho - just FYI and to get another set of eyes

tt-rkim commented on CODEOWNERS:
```diff
@@ -101,6 +101,7 @@ tests/scripts/run_performance.sh @tt-rkim @tapspatel
 tt_eager/tt_dnn/op_library/bmm/ @TT-BrianLiu @bbradelTT
 
 # eager - tensor and op infra
+tt_eager/tt_dnn/op_library/ccl/* @SeanNijjar
```
Delete the `*` - I think github may only match the first level. Need to confirm that though...

tt-rkim commented on tests/tt_eager/module.mk:
```diff
@@ -1,6 +1,7 @@
 # Every variable in subdir must be prefixed with subdir (emulating a namespace)
 TT_EAGER_TESTS += \
 		 tests/tt_eager/ops/ccl/test_all_gather_utils \
+		 tests/tt_eager/ops/ccl/test_all_gather_sharded_indexing_helpers \
```
Hmm probably need to add to cmake?

SeanNijjar replied:
updated

SeanNijjar commented:
Workflows:
post-commit: https://github.com/tenstorrent/tt-metal/actions/runs/9019282375
multichip unit tests: https://github.com/tenstorrent/tt-metal/actions/runs/9019297732

SeanNijjar committed:
#6041: CCL refactoring and sharded allgather optimization

Enable single-tile-high shard width allgather (non-block-based formats) support

SeanNijjar committed:
#6388: Fix race condition that would result in hang in EDM-based multi-chip ops

This hang was caused by a race caused by reuse of a 16B memory region
for two levels of ack sent from receiver to sender side of EDM. The
issue arises because of the following sequence:

EDM Receiver Channel:
1) Receives Payload signal (fine)
2) Sends first level ack to sender
   - Updates `erisc_info->channels[i].receiver_ack` (this becomes a problem)
3) Issues eth send (`src_addr=&erisc_info->channels[i]`)
4) Waits for workers to complet
5) Sends second level ack to sender
   - Updates `erisc_info->channels[i].bytes_sent` to 0 (becomes a problem)
...
Ethernet Subsystem
1) Erisc updates L1 (step 2 above)
2) Gets and buffers first eth send command
3) Erisc updates L1 (step 5 above)
4) Gets and buffers second eth send command
5) Issues first eth send command
   - L1 is already updated to second level ack at this point
   - Sender gets full ack
6) Issues second eth send command
   - Sender gets second (duplicate) full ack
     - It associates that message with the future/next message

To resolve this issue, we force the first level ack to be performed
using a source address for eth_channel_sync_t that doesn't alias
`erisc_info->channels[i]`

Also re-enable all-gather test configs

SeanNijjar commented:
I had to add the fixes for #6388 before I could safely merge this because there was a race condition that caused instability for sharded allgather as well. The fix for #6388 is included in this PR as well.

Closes #6388

tt-rkim commented:
Now that I'm re-looking at this, should we take this as an opportunity to start a new gtest suite for eager, or just leave it as is for now, where the `run_tt_eager.py` script just runs whatever, whether or not the underlying test is gtest or not?

SeanNijjar commented:
> Now that I'm re-looking at this, should we take this as an opportunity to start a new gtest suite for eager, or just leave it as is for now, where the `run_tt_eager.py` script just runs whatever, whether or not the underlying test is gtest or not?

I'm likely going to add more gtests like these over time so I think from an organizational perspective, I think it would make sense to split it off. It's really just testing helpers. If it's alright with you, how about we figure out how to structure it in my next set of changes? I've got some new op coming down the pipe (reduce_scatter) - I can make the switch then?

tt-rkim commented:
Sounds good to me, approved

tt-rkim commented:
I also checked the CPP test run time diff between your branch and main for N300 FD - doesn't seem to be notable
